### DISCUSSION TITLE
Supervision Tree

### DISCUSSION THREAD

**Discussion by winksaville** [AletheiaWareLLC/Wink#20] [General]
I'm totally unfamiliar with Supervision Tree. Can you point me to some documentation and how they work in general and the one you're using as a model.

**Comment by stuartmscott** [comment]
A Supervision Tree enables SMs to monitor other SMs and react when to their lifecycle changes. I first encountered the concept in [Erlang](https://erlang.org/documentation/doc-4.9.1/doc/design_principles/sup_princ.html).

In Wink when a SM spawns another, the parent receives lifecycle messages from the child. See the [family](https://github.com/AletheiaWareLLC/Wink/blob/main/samples/family/parent.cpp) sample.

In the "success" case, the parent will receive;
- started - the child indicates it has started and provides the parent with the name of its binary and the address (ip:port) it has bound to.
- pulsed - the child indicates it is still alive by sending a heartbeat message every 2 seconds.
- exited - the child indicates it has terminated.

In the "failure" case, the parent will receive;
- started - same as above.
- pulsed - same as above.
- errored - the child indicates it has encountered an error by sending the error message to the parent.
- exited - same as above.

If a parent doesn't receive a heartbeat for 10 seconds it will assume the child has failed (maybe the computer crashed, lost power, or the network disconnected).

When a parent is notified that a child has errored, it can chose to do nothing, restart the child, or raise an error. In the last situation, the grandparent will be notified that the parent has errored.

**Comment by winksaville** [comment]
So it's purely a parent/child tree with no centralized supervisor?

**Reply by stuartmscott** [reply]
Correct
